### ISSUE TITLE
should we include a vscode debug config for Jupyterlab dev mode?

### ISSUE THREAD

**Issue by telamonian** [jupyterlab/jupyterlab#8321]
Setting up run/debug config in vscode is very fiddly and manual (and one of the few things that's sooo much easier in intellij). I love the work that @blink1073 has been doing to add vscode debug configs for the test suites in #8319 and in all of his other test modernization PRs. My question is, should we also include a `.vscode/launch.json` at the repo root for debuging Jupyterlab itself in dev mode? I for one still don't know how to set this up (I just launch chrome and then use devtools), so I'd find this very helpful. Ideally we'd add both "debug in chrome" and "debug in firefox" configs.

@saulshanabrook I remember an issue you posted a while back with some related config files. Do you have a vscode debug config for jlab that you know to be currently working?

**Comment by saulshanabrook** [comment]
> @saulshanabrook I remember an issue you posted a while back with some related config files. Do you have a vscode debug config for jlab that you know to be currently working?

No I haven't tried this in a couple of years unfortunately! Would welcome any files.

**Comment by blink1073** [comment]
I don't know off the top of my head how to do this either, but it sounds like a good idea!

**Comment by blois** [comment]
@DonJayamanne do you have any pointers to set this up?

**Comment by saulshanabrook** [comment]
For reference, this was an issue I opened on this a while ago with instructions on how to set it up: https://github.com/jupyterlab/jupyterlab/issues/3979

**Comment by telamonian** [comment]
Here's my best shot so far:

```json
{
            "type": "python",
            "request": "launch",
            "name": "Launch jupyterlab, debug backend/frontend together",

            "args": [
                "--dev-mode",
                "--notebook-dir=${env:HOME}",
                "--no-browser"
            ],
            "module": "jupyterlab.labapp",
            "serverReadyAction": {
                "pattern": "http://localhost:(\\S+)",
                "uriFormat": "http://localhost:%s",
                "action": "debugWithChrome",
                "webRoot": "${workspaceFolder}"
            },
            "envFile": "${env:HOME}/usr/venv/jup-fs/jup-fs.env"
        }
```

This seems to be almost perfect: it starts the notebook server, then launches chrome at the appropriate url + token, and allows for setting breakpoints in python. Unfortunately, breakpoints in the typescript sources don't work. I'll probably need to set some pathing chicanery for the chrome launch, like in @saulshanabrook's example. However, even more unfortunately `"serverReadyAction"` only support the properties I've listed above and not the full set for the proper `"type": "python"` launch job.

It'd be great if we could convince vscode to improve the configurability of `"serverReadyAction"`, but otherwise this approach may be a no-go. The great thing about `"serverReadyAction"` is that it lets us automatically grab the full jupyterlab url + token from the server stdout. It is possible to set up the python and chrome launches as two launches within a compound launch, but then there's no way for the chrome launch to pick up the token. Is there a simple way we can temporarily disable the need for the url token when doing a `--dev-mode` run?

**Comment by blink1073** [comment]
Yeah, there are two notebook app traits, token and disable_xrsf_check iirc (on my phone)

**Comment by telamonian** [comment]
I figured it out! See #8328 

You can actually debug both the python and the typescript sources in one run. You can even go back and forth between breakpoints in the two languages, which is freaking awesome. It does require disabling the token (thanks @blink1073), but I think that's probably okay for a debug dev-mode run

**Comment by DonJayamanne** [comment]
Wow, please could you upload the `launch.json`, 
And perhaps optionally submit a PR.

**Comment by telamonian** [comment]
> Wow, please could you upload the `launch.json`,
> And perhaps optionally submit a PR.

@DonJayamanne ? I don't quite understand. Like I said in the last post, there's already a PR (which includes the `launch.json` in question) at #8328. Or did you mean a PR to a different repo?

In any case, thanks for stopping by and taking a look! If you have any pointers on my `launch.json` or vscode debug in general, I would love to hear them. I ended up spending ~2 days going over the sources of various vscode modules in order to figure out what was possible. Mostly interesting, occasionally very frustrating (eg I wish that `"serverReadyAction"` was [slightly more flexible](https://github.com/microsoft/vscode/blob/ca22de4b17af4ec84394537546998bba195e84cd/extensions/debug-server-ready/package.json#L35)).
